### ISSUE TITLE
Can NNCF quantized model be saved as an ONNX format file?

### ISSUE THREAD

**Issue by sanbuphy** [openvinotoolkit/nncf#1677]
Hello, I think NNCF is a really good quantization tool. However, I have only managed to serialize it into an XML OpenVINO IR file (quantization by default and quantization with controlled accuracy). I was wondering if it is possible to generate an ONNX format file after quantization? Thank you.

**Comment by sanbuphy** [comment]
> @sanbuphy which backend are you using - Torch, TensorFlow, ONNX or OpenVINO - as your source of the model?

Hi! i use onnx, paddle  and openvino

**Comment by kshpv** [comment]
> > @sanbuphy which backend are you using - Torch, TensorFlow, ONNX or OpenVINO - as your source of the model?
> 
> Hi! i use onnx, paddle and openvino

Hello @sanbuphy! Thank you for the question.
If your inference target is ONNXRuntime with OpenVINOExecutionProvider - we propose a quantization API for ONNX models. You can find a sample here - https://github.com/openvinotoolkit/nncf/tree/develop/examples/post_training_quantization/onnx/mobilenet_v2
Moreover, I highly recommend you to read these two pages before starting to work with PostTrainingQuantization API for ONNX - https://github.com/openvinotoolkit/nncf/tree/develop/docs/compression_algorithms/post_training

If you need any help, please, let us know!

**Comment by sanbuphy** [comment]
> > > @sanbuphy which backend are you using - Torch, TensorFlow, ONNX or OpenVINO - as your source of the model?
> > 
> > 
> > Hi! i use onnx, paddle and openvino
> 
> Hello @sanbuphy! Thank you for the question. If your inference target is ONNXRuntime with OpenVINOExecutionProvider - we propose a quantization API for ONNX models. You can find a sample here - https://github.com/openvinotoolkit/nncf/tree/develop/examples/post_training_quantization/onnx/mobilenet_v2 Moreover, I highly recommend you to read these two pages before starting to work with PostTrainingQuantization API for ONNX - https://github.com/openvinotoolkit/nncf/tree/develop/docs/compression_algorithms/post_training
> 
> If you need any help, please, let us know!

Thank you very much! It is very useful for me, but I also want to ask if i can run the quantization model in ONNXRuntime with CPUExecutionProvider instead of  OpenVINOExecutionProvider  ?

**Comment by sanbuphy** [comment]
> > > > > @sanbuphy which backend are you using - Torch, TensorFlow, ONNX or OpenVINO - as your source of the model?
> > > > 
> > > > 
> > > > Hi! i use onnx, paddle and openvino
> > > 
> > > 
> > > Hello @sanbuphy! Thank you for the question. If your inference target is ONNXRuntime with OpenVINOExecutionProvider - we propose a quantization API for ONNX models. You can find a sample here - https://github.com/openvinotoolkit/nncf/tree/develop/examples/post_training_quantization/onnx/mobilenet_v2 Moreover, I highly recommend you to read these two pages before starting to work with PostTrainingQuantization API for ONNX - https://github.com/openvinotoolkit/nncf/tree/develop/docs/compression_algorithms/post_training
> > > If you need any help, please, let us know!
> > 
> > 
> > Thank you very much! It is very useful for me, but I also want to ask if i can run the quantization model in ONNXRuntime with CPUExecutionProvider instead of OpenVINOExecutionProvider ?
> 
> @sanbuphy, you can run the quantized model using ONNXRuntime with CPUExecutionProvider, but you will not get the speedup, in fact, the model could work slower.

Oh! Can I understand it this way, actually it's because of the platform differences in quantized operators? The quantized operator format after NNCF quantization can only be well used in the openvino EP, and there is support for corresponding int8 operators. Even if the model is ONNX mode, the operator is still only well mapped and supported by openvino. So if you want to run on the CPU EP, you can only use ort's own quantization or other quantization tools to make it work?

**Comment by kshpv** [comment]
> > > > > > @sanbuphy which backend are you using - Torch, TensorFlow, ONNX or OpenVINO - as your source of the model?
> > > > > 
> > > > > 
> > > > > Hi! i use onnx, paddle and openvino
> > > > 
> > > > 
> > > > Hello @sanbuphy! Thank you for the question. If your inference target is ONNXRuntime with OpenVINOExecutionProvider - we propose a quantization API for ONNX models. You can find a sample here - https://github.com/openvinotoolkit/nncf/tree/develop/examples/post_training_quantization/onnx/mobilenet_v2 Moreover, I highly recommend you to read these two pages before starting to work with PostTrainingQuantization API for ONNX - https://github.com/openvinotoolkit/nncf/tree/develop/docs/compression_algorithms/post_training
> > > > If you need any help, please, let us know!
> > > 
> > > 
> > > Thank you very much! It is very useful for me, but I also want to ask if i can run the quantization model in ONNXRuntime with CPUExecutionProvider instead of OpenVINOExecutionProvider ?
> > 
> > 
> > @sanbuphy, you can run the quantized model using ONNXRuntime with CPUExecutionProvider, but you will not get the speedup, in fact, the model could work slower.
> 
> Oh! Can I understand it this way: actually it's because of the platform differences in quantized operators? The quantized operator format after NNCF quantization can only be well used in the openvino EP, and there is support for corresponding int8 operators. Even if the model is ONNX mode, the operator is still only well mapped and supported by openvino. So if you want to run on the CPU EP, you can only use ort's own quantization or other quantization tools to make it work?

You are almost correct! I would rephrase this in the following way:
OpenVIMO ExecutionProvider and CPU ExecutionProvider require different quantization schemes. It means that every inference engine expects to have quantizers in specific positions. If there are no quantizers it leads to performance or accuracy issues.
Currently, NNCF does not support a quantization scheme for ONNXRuntime CPU EP.

**Comment by sanbuphy** [comment]
> > > > > > > @sanbuphy which backend are you using - Torch, TensorFlow, ONNX or OpenVINO - as your source of the model?
> > > > > > 
> > > > > > 
> > > > > > Hi! i use onnx, paddle and openvino
> > > > > 
> > > > > 
> > > > > Hello @sanbuphy! Thank you for the question. If your inference target is ONNXRuntime with OpenVINOExecutionProvider - we propose a quantization API for ONNX models. You can find a sample here - https://github.com/openvinotoolkit/nncf/tree/develop/examples/post_training_quantization/onnx/mobilenet_v2 Moreover, I highly recommend you to read these two pages before starting to work with PostTrainingQuantization API for ONNX - https://github.com/openvinotoolkit/nncf/tree/develop/docs/compression_algorithms/post_training
> > > > > If you need any help, please, let us know!
> > > > 
> > > > 
> > > > Thank you very much! It is very useful for me, but I also want to ask if i can run the quantization model in ONNXRuntime with CPUExecutionProvider instead of OpenVINOExecutionProvider ?
> > > 
> > > 
> > > @sanbuphy, you can run the quantized model using ONNXRuntime with CPUExecutionProvider, but you will not get the speedup, in fact, the model could work slower.
> > 
> > 
> > Oh! Can I understand it this way: actually it's because of the platform differences in quantized operators? The quantized operator format after NNCF quantization can only be well used in the openvino EP, and there is support for corresponding int8 operators. Even if the model is ONNX mode, the operator is still only well mapped and supported by openvino. So if you want to run on the CPU EP, you can only use ort's own quantization or other quantization tools to make it work?
> 
> You are almost correct! I would rephrase this in the following way: OpenVINO ExecutionProvider and CPU ExecutionProvider require different quantization schemes. It means that every inference engine expects to have quantizers in specific positions. If quantization placements are unexpected it leads to performance or accuracy issues. Currently, NNCF does not support a quantization scheme for ONNXRuntime CPU EP.

Thank you very much! I completely understand. Here's my final question: Does the openvinoEP of onnxruntime have any impact on multi-threaded tasks? Would using multiple threads (where onnxruntime should be openmp) affect the performance of openvinoEP? After using multiple threads, I found that there is no difference in speed between using openvinoEP and not using it, which is quite amazing.

**Comment by kshpv** [comment]
> > > > > > > > @sanbuphy which backend are you using - Torch, TensorFlow, ONNX or OpenVINO - as your source of the model?
> > > > > > > 
> > > > > > > 
> > > > > > > Hi! i use onnx, paddle and openvino
> > > > > > 
> > > > > > 
> > > > > > Hello @sanbuphy! Thank you for the question. If your inference target is ONNXRuntime with OpenVINOExecutionProvider - we propose a quantization API for ONNX models. You can find a sample here - https://github.com/openvinotoolkit/nncf/tree/develop/examples/post_training_quantization/onnx/mobilenet_v2 Moreover, I highly recommend you to read these two pages before starting to work with PostTrainingQuantization API for ONNX - https://github.com/openvinotoolkit/nncf/tree/develop/docs/compression_algorithms/post_training
> > > > > > If you need any help, please, let us know!
> > > > > 
> > > > > 
> > > > > Thank you very much! It is very useful for me, but I also want to ask if i can run the quantization model in ONNXRuntime with CPUExecutionProvider instead of OpenVINOExecutionProvider ?
> > > > 
> > > > 
> > > > @sanbuphy, you can run the quantized model using ONNXRuntime with CPUExecutionProvider, but you will not get the speedup, in fact, the model could work slower.
> > > 
> > > 
> > > Oh! Can I understand it this way: actually it's because of the platform differences in quantized operators? The quantized operator format after NNCF quantization can only be well used in the openvino EP, and there is support for corresponding int8 operators. Even if the model is ONNX mode, the operator is still only well mapped and supported by openvino. So if you want to run on the CPU EP, you can only use ort's own quantization or other quantization tools to make it work?
> > 
> > 
> > You are almost correct! I would rephrase this in the following way: OpenVINO ExecutionProvider and CPU ExecutionProvider require different quantization schemes. It means that every inference engine expects to have quantizers in specific positions. If quantization placements are unexpected it leads to performance or accuracy issues. Currently, NNCF does not support a quantization scheme for ONNXRuntime CPU EP.
> 
> Thank you very much! I completely understand. Here's my final question: Does the openvinoEP of onnxruntime have any impact on multi-threaded tasks? Would using multiple threads (where onnxruntime should be openmp) affect the performance of openvinoEP? After using multiple threads, I found that there is no difference in speed between using openvinoEP and not using it, which is quite amazing.

I do not have an answer to this question right now. I will try to discover it and back to you or ask my peer to answer.

**Comment by sanbuphy** [comment]
> > > > > > > > > @sanbuphy which backend are you using - Torch, TensorFlow, ONNX or OpenVINO - as your source of the model?
> > > > > > > > 
> > > > > > > > 
> > > > > > > > Hi! i use onnx, paddle and openvino
> > > > > > > 
> > > > > > > 
> > > > > > > Hello @sanbuphy! Thank you for the question. If your inference target is ONNXRuntime with OpenVINOExecutionProvider - we propose a quantization API for ONNX models. You can find a sample here - https://github.com/openvinotoolkit/nncf/tree/develop/examples/post_training_quantization/onnx/mobilenet_v2 Moreover, I highly recommend you to read these two pages before starting to work with PostTrainingQuantization API for ONNX - https://github.com/openvinotoolkit/nncf/tree/develop/docs/compression_algorithms/post_training
> > > > > > > If you need any help, please, let us know!
> > > > > > 
> > > > > > 
> > > > > > Thank you very much! It is very useful for me, but I also want to ask if i can run the quantization model in ONNXRuntime with CPUExecutionProvider instead of OpenVINOExecutionProvider ?
> > > > > 
> > > > > 
> > > > > @sanbuphy, you can run the quantized model using ONNXRuntime with CPUExecutionProvider, but you will not get the speedup, in fact, the model could work slower.
> > > > 
> > > > 
> > > > Oh! Can I understand it this way: actually it's because of the platform differences in quantized operators? The quantized operator format after NNCF quantization can only be well used in the openvino EP, and there is support for corresponding int8 operators. Even if the model is ONNX mode, the operator is still only well mapped and supported by openvino. So if you want to run on the CPU EP, you can only use ort's own quantization or other quantization tools to make it work?
> > > 
> > > 
> > > You are almost correct! I would rephrase this in the following way: OpenVINO ExecutionProvider and CPU ExecutionProvider require different quantization schemes. It means that every inference engine expects to have quantizers in specific positions. If quantization placements are unexpected it leads to performance or accuracy issues. Currently, NNCF does not support a quantization scheme for ONNXRuntime CPU EP.
> > 
> > 
> > Thank you very much! I completely understand. Here's my final question: Does the openvinoEP of onnxruntime have any impact on multi-threaded tasks? Would using multiple threads (where onnxruntime should be openmp) affect the performance of openvinoEP? After using multiple threads, I found that there is no difference in speed between using openvinoEP and not using it, which is quite amazing.
> 
> I do not have an answer to this question right now. I will try to discover it and be back to you shortly or ask my peer to answer.

Thank you very much!

**Comment by saurabhintel0** [comment]
@sanbuphy 
if you are using _options.num_of_threads = 8_ . Which you can see in the (docs here: https://onnxruntime.ai/docs/execution-providers/OpenVINO-ExecutionProvider.html#cc-api)
(options.num_of_threads) It's meaning is different in OVEP context. It actually means to create num_of_infer_requests and have it ready.

By default OVEP creates 8 num_of_threads, which is technically 8 num_of_infer_requests whenever a basic backend is created and of course using the run time flag _--num_of_threads_  you can reduce/increase the number. But 8 is the ideal number is what we thought gives good results across all plugins.

If you want to use multi threading we suggest  you can use it with ONNX Runtime. this is the documentation: https://onnxruntime.ai/docs/performance/tune-performance/threading.html

**Comment by sanbuphy** [comment]
> @sanbuphy if you are using _options.num_of_threads = 8_ . Which you can see in the (docs here: https://onnxruntime.ai/docs/execution-providers/OpenVINO-ExecutionProvider.html#cc-api) (options.num_of_threads) It's meaning is different in OVEP context. It actually means to create num_of_infer_requests and have it ready.
> 
> By default OVEP creates 8 num_of_threads, which is technically 8 num_of_infer_requests whenever a basic backend is created and of course using the run time flag _--num_of_threads_  you can reduce/increase the number. But 8 is the ideal number is what we thought gives good results across all plugins.
> 
> If you want to use multi threading we suggest  you can use it with ONNX Runtime. this is the documentation: https://onnxruntime.ai/docs/performance/tune-performance/threading.html

Thank you for your response, it is helpful for me. I will go and conduct a detailed experiment.
